backend/src/akvo/lumen/component/keycloak.clj

```diff
@@ -220,11 +258,79 @@
           (println (format "Tried to remove user: %s" user-id))
           (lib/internal-server-error))))))
 
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; API Authorization
+;;;
+
+(defn- api-get-async [{:keys [timeout] :as req-opts} url]
+  @(d/timeout!
+    (d/chain
+     (http/get url (select-keys req-opts [:headers :pool]))
+     :body
+     #(stream/map bs/to-byte-array %)
+     #(stream/reduce conj [] %)
+     bs/to-string
+     json/decode)
+    timeout
+    nil))
+
+(defn- active-user [users email]
+  (-> (filter #(and (= (get % "email") email)
+                    (get % "enabled"))
+              users)
+      first
+      (get "id")))
+
+(defn- lookup-user-id
+  "Lookup email -> Keycloak user-id, via cached Keycloak API."
+  [req-opts api-root user-id-cache email]
+  (if-some [user-id (get @user-id-cache email)]
+    user-id
+    (let [url (format "%s/users/?email=%s" api-root email)]
+      (if-some [users (api-get-async req-opts url)]
+        (if-some [user-id (active-user users email)]
+          (-> user-id-cache
+              (swap! assoc email user-id)
+              (get email))
+          (throw
+           (ex-info "No active user with email provided" {:response-code 401})))
+        (throw
+         (ex-info "Down stream auth server timed out (Keycloak)"
+                  {:response-code 503}))))))
+
+(defn allowed-paths
+  "Provided an email address from the authentication process dig out the
+  Keycloak user and get allowed set of paths, as in #{\"demo/admin\" \"t1\"}
+
+  The Keycloak groups are on the form /akvo/lumen/demo/admin, to simplify  we
+  remove the leading /akvo/lumen and return paths as demo/admin."
+  ([keycloak email]
+   (allowed-paths keycloak email {}))
+  ([{:keys [api-root http-pool user-id-cache] :as keycloak} email
+    {:keys [timeout] :or {timeout 10000}}]
+   (let [bare-req-opts {:pool http-pool
+                        :timeout timeout}]
+     (if-some [headers (request-headers keycloak bare-req-opts)]
+       (let [req-opts (assoc bare-req-opts :headers headers)]
+         (if-let [user-id (lookup-user-id req-opts api-root user-id-cache email)]
+           (if-let [allowed-paths (api-get-async req-opts
```

Comment:
This exception handling is copied all over the place. Apart of the duplication, it makes the code harder to follow with so many nested ifs. This code is already non-pure and uses exceptions for flow control (for example `lookup-user-id` throws exceptions and it has no else branch in this logic), so there is really no point on returning nils and handling them

Move to helper function suggested at changes in `api-get-async`

Reply:
Agree on the lack of beauty on this code, this should be reworked.

Reply:
should be solved by https://github.com/akvo/akvo-lumen/pull/2227/commits/1c45ebaaae64f1d83f8d65f9a6114c41c1cb2bd1